File: tron_battle/tron_battle.py
# Copyright (c) 2016 Hubert Jarosz
#
# This software is provided 'as-is', without any express or implied
# warranty. In no event will the authors be held liable for any damages
# arising from the use of this software.
#
# Permission is granted to anyone to use this software for any purpose,
# including commercial applications, and to alter it and redistribute it
# freely, subject to the following restrictions:
#
# 1. The origin of this software must not be misrepresented; you must not
#    claim that you wrote the original software. If you use this software
#    in a product, an acknowledgement in the product documentation would be
#    appreciated but is not required.
# 2. Altered source versions must be plainly marked as such, and must not be
#    misrepresented as being the original software.
# 3. This notice may not be removed or altered from any source distribution.

# currently at gold

import sys  # for sys.stderr
import copy  # for deepcopy
import collections  # for deque
import math  # for sqrt
import operator  # for itemgetter

WIDTH = 30
HEIGHT = 20
MAX_DEPTH = 2

################################################################################

# checks if given coordinate is on map
def is_on_map(x, y):
    if x < 0 or x >= WIDTH or y < 0 or y >= HEIGHT:
        return False
    return True

# gives map field neighbours that are on map
def get_neighbours(x, y):
    first_set = {(x-1, y), (x+1, y), (x, y-1), (x, y+1)}

    return {(vx, vy) for (vx, vy) in first_set if is_on_map(vx, vy)}

# compute distance between two points
def distance(p0, p1):
    return math.sqrt((p0[0] - p1[0])**2 + (p0[1] - p1[1])**2)

#
def getEnemy(players, playerID):
    pl = list(players)
    return pl[0] if pl[0] != playerID else pl[1]

################################################################################

class GameMap:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.map = [[-1]*self.height for n in range(self.width)]
        #self.map = [[-1] * self.height] * self.width   ## BAD!!!!
    def set(self, x, y, val=0):
        self.map[x][y] = val
    def get(self, x, y):
        return self.map[x][y]

    def freeNeighbours(self, x, y):
        neibs = get_neighbours(x,y)
        fn = [ (nx,ny) for (nx,ny) in neibs if self.get(nx,ny) < 0 ]
        return fn

    def clone(self):
        return copy.deepcopy(self)

    def __moveWest(self, w, y):
        TARGET = -1
        while w >= 0 and self.map[w][y] == TARGET:
            w -= 1
        return w+1

    def __moveEast(self, e, y):
        TARGET = -1
        while e < WIDTH and self.map[e][y] == TARGET:
            e += 1
        return e ## no -1 because we use it in range() - dirty speedhack

    def expand(self, start):

        ID = self.get(*start)
        fn = self.freeNeighbours(*start)
        for x, y in fn:
            self.set(x, y, ID)
        return set(fn)

    def fill(self, playerID, players):

        startingPositions = {p: [coord] for p, coord in players.items()}
        gameMap = self.clone()

        playerList = list(players)
        data = {pid: {} for pid in playerList}

        playerOrder = [playerID] + [p for p in playerList if p > playerID] + [p for p in playerList if p < playerID]

        turnNo = 1
        canExplore = True

        while canExplore:
            canExplore = False
            movesThisTurn = {}
            for player in playerOrder:
                for pos in startingPositions[player]:
                    for n in gameMap.freeNeighbours(*pos):
                        canExplore = True
                        gameMap.map[pos[0]][pos[1]] = player
                        movesThisTurn[n] = player
            #
            for move, player in movesThisTurn.items():
                data[player][move] = turnNo

            if canExplore:
                turnNo += 1
                startingPositions = {player: [move for move, p in movesThisTurn.items() if p == player] for player in playerList}
        #

        playerTiles = len(data[playerID])
        enemyTiles = sum([len(data[i]) for i in playerList]) - playerTiles

        return (playerTiles, enemyTiles)


    def clean(self, pid):
        for x in range(len(self.map)):
            for y in range(len(self.map[x])):
                if self.map[x][y] == pid:
                    self.map[x][y] = -1



################################################################################


def moveScore(gM, playerID, players):
    p, e = gM.fill(playerID, players)
    score = 10000 * p - 10 * e - len(gM.freeNeighbours(*players[playerID])) # + acc

    #print(players[playerID], p, file=sys.stderr)
    return score


# def minmax(gameMap, depth, playerID, maximizing, players, alpha, beta):
#
#     enemyID = getEnemy(players, playerID)
#     #enemyID = 0 if playerID > 0 else 1
#
#     fn = gameMap.freeNeighbours(*players[playerID])
#
#     if depth == 0 or len(fn) == 0:
#         x, y = players[playerID]
#         return moveScore(gameMap, playerID, players)
#
#     if maximizing:
#         v = []
#         for n_x, n_y in fn:
#             gM = gameMap.clone()
#             p = copy.deepcopy(players)
#             gM.set(n_x, n_y, playerID)
#             p[playerID] = (n_x, n_y)
#             alpha = max(alpha, minmax(gM, depth - 1, playerID, True, p, alpha, beta))
#             if alpha >= beta:
#                 break # cut beta
#         return alpha
    # else:
    #     x, y = players[enemyID]
    #     fn = gameMap.freeNeighbours(x, y)
    #
    #     v = []
    #     for n_x, n_y in fn:
    #         gM = gameMap.clone()
    #         p = copy.deepcopy(players)
    #         gM.set(n_x, n_y, enemyID)
    #         p[enemyID] = (n_x, n_y)
    #         beta = min(beta, minmax(gM, depth - 1, playerID, True, p, alpha, beta))
    #         if alpha >= beta:
    #             break # cut alpha
    #     return beta

################################################################################


def playField(gameMap, playerID, players):
    #print("start:", players[playerID], file=sys.stderr)
    fn = gameMap.freeNeighbours(*players[playerID])
    #print("fn:", fn, file=sys.stderr)

    if len(fn) == 0:
        print("Dead.", file=sys.stderr)
        return (-1, -1)

    maxV = 0
    n = fn[0]
    alpha, beta = float("-inf"), float("inf")
    for n_x, n_y in fn:
        gM = gameMap.clone()
        p = copy.deepcopy(players)
        gM.set(n_x, n_y, playerID)

        p[playerID] = (n_x, n_y)

        v = moveScore(gM, playerID, p)

        #v = minmax(gM, MAX_DEPTH, playerID, True, p , alpha, beta)

        if v > maxV:
            maxV = v
            n = (n_x, n_y)
    return n


################################################################################

gameMap = GameMap(WIDTH, HEIGHT)

stillInGame = None
players = dict()

firstTurn = True

# game loop
while True:
    # n: total number of players (2 to 4).
    # p: your player number (0 to 3).
    n, p = [int(i) for i in input().split()]

    if stillInGame == None:
        stillInGame = [True] * n

    x, y = -1, -1
    nbr = 0
    enX, enY = -1, -1
    for i in range(n):

        x0, y0, x1, y1 = [int(j) for j in input().split()]

        if firstTurn:
            gameMap.set(x0, y0, i)

        if x1 != -1:
            players[i] = (x1, y1)
            gameMap.set(x1, y1, i)
        else:
            if stillInGame[i]:
                gameMap.clean(i)
                stillInGame[i] = False
                del players[i]

    firstTurn = False

    x, y = players[p]

    if (x,y) == (-1,-1):
        print("I'm dead.", file=sys.stderr)
        break


    go_to = playField(gameMap, p, players)
    print((x,y),"->", go_to, file=sys.stderr)

    if go_to[0] > x:
        print("RIGHT")
    elif go_to[0] < x:
        print("LEFT")
    elif go_to[1] > y:
        print("DOWN")
    else:
        print("UP")
    # To debug: print("Debug messages...", file=sys.stderr)

File: tron_battle/tron_battle.go
package main

import (
	"fmt"
	"math"
	"os"
	"sort"
)

const (
	WIDTH     = 30
	HEIGHT    = 20
	MAX_DEPTH = 2
	FREE      = -1
)

var myID int = -1

const maxInt = int(^uint(0) >> 1)
const minInt = -maxInt - 1

func max(a int, b int) int {
	if a < b {
		return b
	}
	return a
}

func min(a int, b int) int {
	if a > b {
		return b
	}
	return a
}

type gameMap_t [WIDTH][HEIGHT]int
type pCoords map[int]coordinate

type coordinate struct {
	x int
	y int
}

func dirToStr(from coordinate, to coordinate) string {
	switch {
	case from.x < to.x && from.y == to.y:
		return "RIGHT"
	case from.x > to.x && from.y == to.y:
		return "LEFT"
	case from.y < to.y && from.x == to.x:
		return "DOWN" // inverted y
	case from.y > to.y && from.x == to.x:
		return "UP" // inverted y
	}
	return "INVALID"
}

func getNeighbours(coord coordinate) []coordinate {
	neighbours := make([]coordinate, 0, 4)
	if coord.x-1 >= 0 {
		neighbours = append(neighbours, coordinate{coord.x - 1, coord.y})
	}
	if coord.x+1 < WIDTH {
		neighbours = append(neighbours, coordinate{coord.x + 1, coord.y})
	}
	if coord.y-1 >= 0 {
		neighbours = append(neighbours, coordinate{coord.x, coord.y - 1})
	}
	if coord.y+1 < HEIGHT {
		neighbours = append(neighbours, coordinate{coord.x, coord.y + 1})
	}
	return neighbours
}

func (g gameMap_t) getFreeNeighbours(coord coordinate) []coordinate {
	freeNeighbours := make([]coordinate, 0, 4)
	for _, n := range getNeighbours(coord) {
		if g.get(n) == FREE {
			freeNeighbours = append(freeNeighbours, n)
		}
	}
	return freeNeighbours
}

func (g gameMap_t) get(coord coordinate) int {
	return g[coord.x][coord.y]
}

func (g *gameMap_t) set(coord coordinate, val int) {
	g[coord.x][coord.y] = val
}

func (g *gameMap_t) boardValue(players pCoords) (int, bool) {

	wallhug := len(g.getFreeNeighbours(players[myID]))
	p, e, alone := g.getValues(players)
	score := 10000*p - 10*e - wallhug

	//fmt.Fprintln(os.Stderr, players[myID], p, e, alone)

	return score, alone
}

func (g *gameMap_t) preparePlayers(players pCoords) ([]int, map[int][]coordinate) {
	startingPositions := make(map[int][]coordinate)

	pre := make([]int, 0, 3)
	order := make([]int, 0, 4)

	for p, pos := range players {
		if p < myID {
			pre = append(pre, p)
		}
		if p > myID {
			order = append(order, p)
		}

		startingPositions[p] = append(startingPositions[p], pos)
	}
	sort.Ints(pre)
	sort.Ints(order)
	order = append(order, pre...)
	order = append(order, myID)

	return order, startingPositions
}

func (g gameMap_t) getValues(players pCoords) (int, int, bool) {
	var order []int
	var positions map[int][]coordinate
	order, positions = g.preparePlayers(players)

	pFields := 0
	eFields := 0
	canExplore := true
	alone := true

	for canExplore {
		canExplore = false
		movesThisTurn := make(map[coordinate]int)

		for _, player := range order {
			for _, coord := range positions[player] {
				neighbours := getNeighbours(coord)
				for _, n := range neighbours {
					value := g.get(n)
					if value == FREE {
						canExplore = true
						g.set(n, player+10)
						movesThisTurn[n] = player
					} else if value >= 10 && value != myID+10 && player == myID {
						alone = false
					}
				}
			}
		}

		positions = make(map[int][]coordinate)

		for move, player := range movesThisTurn {
			if player == myID {
				pFields++
			} else {
				eFields++
			}
			positions[player] = append(positions[player], move)
		}
	}
	return pFields, eFields, alone
}

func (p pCoords) getEnemy() int {

	d := (WIDTH + HEIGHT) * 10.0
	enemyID := -1

	for id, coord := range p {
		if id != myID {
			dist := distance(p[myID], coord)
			if dist < d {
				d = dist
				enemyID = id
			}
		}
	}

	return enemyID
}

func getNextMove(g gameMap_t, players pCoords) coordinate {
	nextMove := coordinate{-1, -1}

	myPos := players[myID]
	fn := g.getFreeNeighbours(myPos)
	if len(fn) <= 0 {
		fmt.Fprintln(os.Stderr, "No moves :(")
		return nextMove
	}

	maxV := minInt
	_, alone := g.boardValue(players)

	first := true

	for _, n := range fn {
		g.set(n, myID)
		players[myID] = n

		var v int

		if alone {
			v, _ = g.boardValue(players)
		} else {
			v = minmax(g, players, MAX_DEPTH, false, minInt, maxInt)
		}

		if v > maxV || first {
			maxV = v
			nextMove = n
		}

		g.set(n, FREE)
		first = false
	}

	players[myID] = myPos

	return nextMove
}

func (g *gameMap_t) init() {
	for x := 0; x < WIDTH; x++ {
		for y := 0; y < HEIGHT; y++ {
			g[x][y] = FREE
		}
	}
}

func (g *gameMap_t) clean(id int) {
	for x := 0; x < WIDTH; x++ {
		for y := 0; y < HEIGHT; y++ {
			if g[x][y] == id {
				g[x][y] = FREE
			}
		}
	}
}

func distance(c0, c1 coordinate) float64 {
	x := c0.x - c0.x
	y := c0.y - c1.y
	return math.Sqrt(float64(x*x + y*y))
}

func main() {

	players := make(pCoords)
	stillInGame := [4]bool{true, true, true, true}
	var gameMap gameMap_t
	gameMap.init()

	for {
		// N: total number of players (2 to 4).
		// P: your player number (0 to 3).
		var N, P int
		fmt.Scan(&N, &P)
		var starting, current coordinate

		for i := 0; i < N; i++ {
			// X0: starting X coordinate of lightcycle (or -1)
			// Y0: starting Y coordinate of lightcycle (or -1)
			// X1: current X coordinate of lightcycle
			// Y1: current Y coordinate of lightcycle

			fmt.Scan(&starting.x, &starting.y, &current.x, &current.y)

			if myID < 0 {
				gameMap.set(starting, i)
			}

			if current.x == -1 {
				if stillInGame[i] {
					stillInGame[i] = false
					gameMap.clean(i)
					delete(players, i)
				}
			} else {
				players[i] = current
				gameMap.set(current, i)
			}

		}

		if myID < 0 {
			myID = P
		}

		nextMove := getNextMove(gameMap, players)

		fmt.Fprintln(os.Stderr, players[myID], "->", nextMove)
		fmt.Println(dirToStr(players[myID], nextMove))
	}
}

func minmax(g gameMap_t, players pCoords, depth int, maximizing bool, alpha int, beta int) int {
	fn := g.getFreeNeighbours(players[myID])

	if depth == 0 || len(fn) <= 0 {
		v, _ := g.boardValue(players)
		return v
	}

	if maximizing {
		for _, n := range fn {
			p := players.clone()
			g.set(n, myID)
			p[myID] = n

			alpha = max(alpha, minmax(g, p, depth-1, false, alpha, beta))
			g.set(n, FREE)

			if alpha >= beta {
				break
			}
		}
		return alpha
	} else {

		for enemyID, coord := range players {
			if enemyID != myID {
				fn = g.getFreeNeighbours(coord)
				for _, n := range fn {
					p := players.clone()

					g.set(n, enemyID)
					p[enemyID] = n

					beta = min(beta, minmax(g, p, depth-1, true, alpha, beta))
					g.set(n, FREE)

					if alpha >= beta {
						return beta
					}
				}
			}
		}
		return beta
	}
}

func (p pCoords) clone() pCoords {
	pClone := make(pCoords)
	for k, v := range p {
		pClone[k] = v
	}
	return pClone
}
